File: android/gemcore/src/main/kotlin/com/gemwallet/android/model/ValueFormatter.kt
package com.gemwallet.android.model

import android.icu.text.CompactDecimalFormat
import com.wallet.core.primitives.Asset
import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.Locale

class ValueFormatter(
    private val style: Style,
    private val locale: Locale = Locale.getDefault(),
    private val abbreviationThreshold: BigDecimal = ABBREVIATION_THRESHOLD,
) {
    enum class Style { Short, Compact, Auto, Full }

    fun string(value: BigInteger, asset: Asset): String =
        string(value, decimals = asset.decimals, currency = asset.symbol)

    fun string(value: BigInteger, decimals: Int, currency: String = ""): String =
        string(BigDecimal(value).movePointLeft(decimals), currency)

    fun string(value: BigDecimal, currency: String = ""): String {
        if (value.signum() == 0) return appendCurrency("0", currency)

        if (style == Style.Short && value.abs() >= abbreviationThreshold) {
            return appendCurrency(abbreviated(value), currency)
        }

        val formatter = (NumberFormat.getInstance(locale) as DecimalFormat).apply {
            roundingMode = RoundingMode.DOWN
        }
        return appendCurrency(formatter.format(value, precision(value.abs())), currency)
    }

    private fun precision(magnitude: BigDecimal): Precision = when (style) {
        Style.Full -> Precision.full
        Style.Short -> if (magnitude >= SMALL_AMOUNT_THRESHOLD) Precision.upToTwoPlaces else Precision.upToFourPlaces
        Style.Compact -> if (magnitude >= SMALL_AMOUNT_THRESHOLD) Precision.upToTwoPlaces else Precision.upToFourPlaces
        Style.Auto -> when {
            magnitude >= BigDecimal.ONE -> Precision.upToTwoPlaces
            magnitude >= DUST_THRESHOLD -> Precision.fourSignificant
            else -> Precision.full
        }
    }

    private fun abbreviated(decimal: BigDecimal): String {
        val formatter = CompactDecimalFormat.getInstance(locale, CompactDecimalFormat.CompactStyle.SHORT)
        formatter.maximumFractionDigits = 2
        return formatter.format(decimal)
    }

    private fun appendCurrency(value: String, currency: String): String =
        if (currency.isEmpty()) value else "$value $currency"

    companion object {
        private val SMALL_AMOUNT_THRESHOLD: BigDecimal = BigDecimal("0.1")
        private val DUST_THRESHOLD: BigDecimal = BigDecimal("0.0001")
        val ABBREVIATION_THRESHOLD: BigDecimal = BigDecimal(10_000)
    }
}

File: ios/Packages/Formatters/Sources/ValueFormatter.swift
// Copyright (c). Gem Wallet. All rights reserved.

import BigInt
import Foundation
import Primitives

public struct ValueFormatter: Sendable {
    public enum Style: Sendable {
        case full, short, auto
    }

    private static let smallAmountThreshold = Decimal(sign: .plus, exponent: -1, significand: 1)
    private static let dustThreshold = Decimal(sign: .plus, exponent: -4, significand: 1)

    private let locale: Locale
    private let style: Style
    private let abbreviationThreshold: Decimal

    public init(
        locale: Locale = .current,
        style: Style,
        abbreviationThreshold: Decimal = defaultAbbreviationThreshold,
    ) {
        self.locale = locale
        self.style = style
        self.abbreviationThreshold = abbreviationThreshold
    }

    public func string(_ value: BigInt, asset: Asset) -> String {
        string(value, decimals: asset.decimals.asInt, currency: asset.symbol)
    }

    public func string(_ value: BigInt, decimals: Int, currency: String = "") -> String {
        guard let decimal = BigNumberFormatter.standard.decimal(from: value, decimals: decimals) else {
            return ""
        }
        if value.isZero {
            return appendingCurrency("0", currency: currency)
        }
        if style == .short, abs(decimal) >= abbreviationThreshold, let abbreviated = abbreviatedFormatter.string(from: decimal) {
            return appendingCurrency(abbreviated, currency: currency)
        }
        if style == .short, abs(decimal) < Self.dustThreshold {
            return appendingCurrency("<\(formattedDustThreshold)", currency: currency)
        }
        return appendingCurrency(decimal.formatted(formatStyle(for: decimal)), currency: currency)
    }

    public func inputNumber(from string: String, decimals: Int) throws -> BigInt {
        try BigNumberFormatter.standard.number(
            from: NumberInputNormalizer.normalize(string, locale: locale),
            decimals: decimals,
        )
    }

    public func double(from number: BigInt, decimals: Int) throws -> Double {
        guard let result = BigNumberFormatter.standard.double(from: number, decimals: decimals) else {
            throw AnyError("unknown \(number) number")
        }
        return result
    }

    func number(amount: String) throws -> Decimal {
        try number(amount: amount, locale: locale)
    }

    func number(amount: String, locale: Locale) throws -> Decimal {
        guard let decimal = Decimal(string: amount, locale: locale) else {
            throw AnyError("unknown \(amount) decimal")
        }
        return decimal
    }
}

private extension ValueFormatter {
    var abbreviatedFormatter: AbbreviatedFormatter {
        AbbreviatedFormatter(locale: locale, threshold: abbreviationThreshold)
    }

    var formattedDustThreshold: String {
        Self.dustThreshold.formatted(
            Decimal.FormatStyle()
                .locale(locale)
                .precision(.fractionLength(4)),
        )
    }

    func formatStyle(for decimal: Decimal) -> Decimal.FormatStyle {
        Decimal.FormatStyle()
            .locale(locale)
            .grouping(.automatic)
            .rounded(rule: .towardZero)
            .precision(precision(for: abs(decimal)))
    }

    func precision(for magnitude: Decimal) -> NumberFormatStyleConfiguration.Precision {
        switch (style, magnitude) {
        case (.full, _): .full
        case (.short, Self.smallAmountThreshold...): .upToTwoPlaces
        case (.short, _): .upToFourPlaces
        case (.auto, 1...): .upToTwoPlaces
        case (.auto, Self.dustThreshold...): .fourSignificant
        case (.auto, _): .full
        }
    }

    func appendingCurrency(_ value: String, currency: String) -> String {
        currency.isEmpty ? value : "\(value) \(currency)"
    }
}
